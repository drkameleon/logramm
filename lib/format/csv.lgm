/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/csv.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;
import core.string;

//================================================
// Module definition
//================================================

Csv :: 
{
	//================================================
	// Functions
	//================================================

	parse(csv) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes csv string and returns corresponding object"
	] => 
	{
		lines = csv.split("\n");
		lines = lines.map("trim($)").filter("$!=\"\"");

		header = lines[0];
		headers = header.split(",").map("trim($)");

		ret = [];

		@ line->lines : 
		{
			line!=lines.first() : {

				parts = line.split(",").map("trim($)").filter("$!=\"\"");

				(count(parts)==count(headers)) :
				{
					row = [:];

					@ i,part->parts : row += [ headers[i] : part ];

					ret += row;
				}
			}
		}

		==> ret;
	}

	generate(object) # [
		.params		: "array",
		.returns	: "string",
		.help		: "Processes input and returns corresponding Json as string"
	] => 
	{
		ret = "";

		headers = object.first().keys().map("trim($)");
		
		ret += headers.join(",") + "\n";

		@ row->object : ret += row.values().join(",") + "\n";

		==> ret;
	}
	
}

