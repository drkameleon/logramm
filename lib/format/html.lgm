/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/html.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;

import core.string;

//================================================
// Module definition
//================================================

Html :: 
{

	//================================================
	// Functions
	//================================================

	external parse(markup) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Converts HTML/LGMX template to HTML and returns result"
	] => "html.parse";
	
}
