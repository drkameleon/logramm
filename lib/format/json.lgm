/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/json.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Json :: 
{

	//================================================
	// External D-bindings
	//================================================

	external parse(json) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes json string and returns corresponding object"
	] => "json.parse";

	external generate(object) # [
		.params		: "any",
		.returns	: "string",
		.help		: "Processes input and returns corresponding Json as string"
	] => "json.generate";

}

