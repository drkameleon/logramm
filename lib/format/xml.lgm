/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/xml.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

//================================================
// Module definition
//================================================

Xml :: 
{

	//================================================
	// External D-bindings
	//================================================

	external check(xml) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks an XML document, as string, for well-formedness and returns true if successful"
	] => "xml.check";

}

