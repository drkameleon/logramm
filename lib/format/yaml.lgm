/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/yaml.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Yaml :: 
{

	//================================================
	// External D-bindings
	//================================================

	external parse(yaml) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes YAML string and returns corresponding object"
	] => "yaml.parse";

	external generate(object) # [
		.params		: "any",
		.returns	: "string",
		.help		: "Processes input and returns corresponding YAML as string"
	] => "yaml.generate";

}

