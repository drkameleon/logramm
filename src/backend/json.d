/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** src/backend/json.d
 **********************************************************/

module backend.json;

//================================================
// Imports
//================================================

import std.array;
import std.conv;
import std.json;
import std.stdio;
import std.string;

import value;

//================================================
// Functions
//================================================

Value parseJsonNode(JSONValue n)
{
	switch (n.type())
	{
		case JSON_TYPE.STRING 		: return new Value(n.str());
		case JSON_TYPE.INTEGER 		: return new Value(n.integer());
		case JSON_TYPE.UINTEGER 	: return new Value(n.uinteger());
		case JSON_TYPE.FLOAT 		: return new Value(n.floating());
		case JSON_TYPE.ARRAY 		: {
			Value[] ret;

			foreach (JSONValue v; n.array())
			{
				ret ~= parseJsonNode(v);
			}

			return new Value(ret);
		}
		case JSON_TYPE.OBJECT 		: {
			Value[Value] ret;

			foreach (string k, JSONValue v; n.object())
			{
				Value key = new Value(k);
				Value val = parseJsonNode(v);

				ret[key] = val;
			}

			return new Value(ret);
		}
		default : break;

	}

	return new Value(0); // Shouldn't ever reach here
}

JSONValue generateJsonValue(Value input)
{
	JSONValue ret;
	switch (input.type)
	{
		case ValueType.numberValue		:	ret = input.content.i; return ret;
		case ValueType.realValue		:	ret = input.content.r; return ret;
		case ValueType.stringValue		:	ret = input.content.s; return ret;
		case ValueType.booleanValue		:	ret = to!int(input.content.b); return ret;
		case ValueType.arrayValue		:	{
			JSONValue[] result;
			for (int i=0; i<input.content.a.length; i++)
			{
				result ~= generateJsonValue(input.content.a[i]);
			}
			ret = result;
			return ret;
		}
		case ValueType.dictionaryValue	: 	{
			JSONValue[string] result;
			foreach (Value kv, Value vv; input.content.d)
			{
				result [kv.str()] = generateJsonValue(vv);
			}
			ret = result;
			return ret;
		}
		default	: break;
	}

	return ret; // won't reach here
}

//================================================
// Class
//================================================

class LGM_Json
{
	static Value parse(Value[] v)
	{
		string str = v[0].content.s;

		JSONValue j = parseJSON(str);
		Value ret = parseJsonNode(j);

		return new Value(ret);
	}

	static Value generate(Value[] v)
	{
		Value arg = v[0];

		JSONValue j = generateJsonValue(arg);
		string ret = j.toString();

		return new Value(ret);
	}

}
