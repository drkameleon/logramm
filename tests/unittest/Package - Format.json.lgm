import format.json;

users =
[
	[.username:"drkameleon", .password:"1234567", .age:28],
	[.username:"ozmax",		 .password:"7654321", .age:24]
];

json = generate(users);

out json;
out "==============";
out parse(json);