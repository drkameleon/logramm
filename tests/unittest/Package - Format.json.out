[{"username":"drkameleon","password":"1234567","age":28},{"username":"ozmax","password":"7654321","age":24}]
==============
[["username":"drkameleon", "password":"1234567", "age":28],["username":"ozmax", "password":"7654321", "age":24]]
